### PR TITLE
Publish or put project into draft

### PR DESCRIPTION
**Pull request by temi**

### CODE REVIEW

**Review by temi — COMMENTED** [review]
Please see my comments. Make changes to ecodata `Project` domain object to add appropriate fields. Another thing missing is making sure draft projects do not appear in project finder. Happy to discuss on any comments further.

File: grails-app/views/project/create.gsp [hunk 1]
```diff
@@ -127,6 +129,53 @@ $(function(){
             } else {
                 var projectErrors = viewModel.transients.projectHasErrors()
                 if (!projectErrors) {
+                    viewModel.projLifecycleStatus = 'Draft';
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/views/project/create.gsp:132]: Discussed and not valid.

File: grails-app/views/project/edit.gsp [hunk 2]
```diff
@@ -56,6 +56,7 @@
     <div class="row">
         <div class="col-12 btn-space">
             <button type="button" id="save" class="btn btn-primary-dark"><i class="fas fa-hdd"></i> <g:message code="g.save"/></button>
+            <button type="button" id="publish" class="btn btn-primary-dark" data-bind="enable: (isExternal() || isWorks() || (!isExternal() && transients.hasPublishedActivities)), text:publishUnpublish()"><i class="fas fa-hdd"></i></button>
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/views/project/edit.gsp:59]: Discussed and not valid.

File: grails-app/views/project/create.gsp [hunk 3]
```diff
@@ -64,7 +64,9 @@
         <div class="row" style="display: none" data-bind="visible: true">
             <div class="col-12 btn-space">
                 <div class="alert warning" data-bind="visible: !termsOfUseAccepted() && !isExternal()"><g:message code="project.details.termsOfUseAgreement.saveButtonWarning"/></div>
-                <button type="button" id="save" class="btn btn-primary-dark" data-bind="disable: (!termsOfUseAccepted() && !isExternal())"><i class="fas fa-hdd"></i> <g:message code="g.save"/></button>
+
+                <button type="button" id="save" class="btn btn-primary-dark" data-bind="disable: (!termsOfUseAccepted() && !isExternal())" title="<g:message code="g.save.title"/>"><i class="fas fa-hdd"></i> <g:message code="g.save"/></button>
+                <button type="button" id="publish" class="btn btn-primary-dark" data-bind="enable: (isExternal() || (isWorks() && termsOfUseAccepted()))"><i class="fas fa-hdd"></i> <g:message code="g.savePublish"/></button>
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/views/project/create.gsp:69]: added comment in code.

File: grails-app/views/project/edit.gsp (hunk 2 above)
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/views/project/edit.gsp:59]: Discussed, added comment in code for further clarity.

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy [hunk 4]
```diff
@@ -484,6 +488,30 @@ class ProjectController {
         String mainImageAttribution = values.remove("mainImageAttribution")
         String logoAttribution = values.remove("logoAttribution")
 
+        def hasPublishedActivities = false
+
+        if (project != null) {
+            def activities = projectActivityService.getAllByProject(project.projectId, "docs", params?.version, true)
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy:494]: done

File: grails-app/assets/javascripts/projects.js [hunk 5]
```diff
@@ -980,6 +981,26 @@ function ProjectViewModel(project, isUserEditor) {
         'Other'
     ];
 
+    self.transients.hasPublishedActivities = false;
+
+    self.checkPublishedActivities = function (activities) {
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/assets/javascripts/projects.js:986]: done

File: grails-app/assets/javascripts/projects.js [hunk 6]
```diff
@@ -312,6 +312,7 @@ function ProjectViewModel(project, isUserEditor) {
         isUserEditor = false;
     }
 
+    self.projLifecycleStatus = 'Draft';
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/assets/javascripts/projects.js:315]: done

File: grails-app/assets/javascripts/projects.js [hunk 7]
```diff
@@ -980,6 +981,26 @@ function ProjectViewModel(project, isUserEditor) {
         'Other'
     ];
 
+    self.transients.hasPublishedActivities = false;
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/assets/javascripts/projects.js:984]: done

File: grails-app/assets/javascripts/projects.js [hunk 8]
```diff
@@ -980,6 +981,26 @@ function ProjectViewModel(project, isUserEditor) {
...
+                if (activities[i].published) {
+                    self.transients.hasPublishedActivities = true;
+                    break;
+                }
+            }
+        }
+    };
+
+    self.publishUnpublish = function () {
+        if (project.projLifecycleStatus == 'Draft')
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/assets/javascripts/projects.js:998]: done

File: grails-app/controllers/au/org/ala/biocollect/ProjectActivityController.groovy [hunk 9]
```diff
@@ -46,6 +46,11 @@ class ProjectActivityController {
         def project = projectService.get(pActivity?.projectId)
         String userId = userService.getCurrentUserId()
 
+        //publish the project along with the survey
+        if (values.publishProject == "Yes") {
+            projectService.update(project?.projectId, [projLifecycleStatus: 'Published'])
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/controllers/au/org/ala/biocollect/ProjectActivityController.groovy:51]: done

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy [hunk 10]
```diff
@@ -302,14 +302,16 @@ class ProjectController {
 
         if (project) {
             def siteInfo = siteService.getRaw(project.projectSiteId)
-            [project: project,
-             siteDocuments: siteInfo.documents?:'[]',
-             site: siteInfo.site,
-             programs: metadataService.programsModel(),
-             scienceTypes: scienceTypes,
-             ecoScienceTypes: ecoScienceTypes
+            def activities = projectActivityService.getAllByProject(project.projectId, "docs", params?.version, true)
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy:305]: done

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy [hunk 11]
```diff
@@ -353,6 +355,8 @@ class ProjectController {
             project.projectType = ProjectService.PROJECT_TYPE_ECOSCIENCE
         }
 
+        project.projLifecycleStatus = 'Draft'
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy:358]: done

File: grails-app/views/project/create.gsp (hunk 1 above)
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/views/project/create.gsp:132]: Discussed and invalid, 'Draft' renamed to unpublished.

File: grails-app/views/project/edit.gsp [hunk 12]
```diff
@@ -125,6 +128,51 @@ $(function(){
...
+
+    $('#publish').click(function () {
+    if(viewModel.transients.kindOfProject() == 'citizenScience' && !viewModel.transients.isDataEntryValid()){
+        bootbox.dialog({message:"Use of this system for data collection is not available for non-biodiversity related projects." +
+            " Press continue to turn data collection feature off. Otherwise, press cancel to modify the form."}, [{
+              label: "Continue",
+              className: "btn-primary",
+              callback: function() {
+                viewModel.isExternal(true);
+                $('#save').click()
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/views/project/edit.gsp:140]: done

**Pull request closed without merging**